flink-core/src/main/java/org/apache/flink/core/memory/DirectMemorySegment.java

```
@@ -0,0 +1,560 @@
+ *  
+ * In addition, the code that uses this class should make sure that only one implementation class is ever loaded -
+ * Either the {@link HeapMemorySegment}, or this DirectMemorySegment. That way, all the abstract methods in the
+ * MemorySegment base class have only one loaded actual implementation. This is easy for the JIT to recognize through
+ * class hierarchy analysis, or by identifying that the invocations are monomorphic (all go to the same concrete
+ * method implementation). Under this precondition, the JIT can perfectly inline methods.
+ * 
+ * This is harder to do and control with byte buffers, where different code paths use different versions of the class
+ * (heap, direct, mapped) and thus virtual method invocations are polymorphic and are not as easily inlined.
+ */
+public class DirectMemorySegment extends MemorySegment {
+	
+	/** The direct byte buffer that allocated the memory */
+	protected final ByteBuffer buffer;
+	
+	/** The address to the off-heap data */
+	private long address;
+	
+	/** The address one byte after the last addressable byte.
+	 *  This is address + size while the segment is not disposed */
+	private final long addressLimit;
+
+	/** The size in bytes of the memory segment */
+	private final int size;
+	
+	// -------------------------------------------------------------------------
+	//                             Constructors
+	// -------------------------------------------------------------------------
+
+	public DirectMemorySegment(int size) {
+		this(ByteBuffer.allocateDirect(size));
+	}
+
+	public DirectMemorySegment(ByteBuffer buffer) {
+		if (buffer == null || !buffer.isDirect()) {
+			throw new IllegalArgumentException();
+		}
+		
+		this.buffer = buffer;
+		this.size = buffer.capacity();
+		this.address = getAddress(buffer);
+		this.addressLimit = this.address + size;
+		
+		if (address >= Long.MAX_VALUE - Integer.MAX_VALUE) {
+			throw new RuntimeException("Segment initialized with too large address: " + address);
```

It's good to fail fast :), but how could this happen? We don't have any control over the address.